compiler/noirc_frontend/src/elaborator/mod.rs

```diff
@@ -1692,4 +1709,78 @@ impl<'context> Elaborator<'context> {
                 _ => true,
             })
     }
+
+    /// Check that a recursive function *can* return without endlessly calling itself.
+    fn check_for_unbounded_recursion(
```

Comment:
We have a `lints.rs` module which this could go in.

Reply:
I think the reason I didn't put it in there was because all lints are arranged under the `fn run_function_lints(&mut self, func: &FuncMeta, modifiers: &FunctionModifiers) {` function which didn't have all the inputs needed in order to call this check, notably it doesn't have the `body_id: ExprId` parameter and the `hir_func` I passed isn't available yet. I suppose after line 499 it will somehow be associated with the function definition, but it felt like lints were intended to do something else. So I followed the example of [type_check_function_body](https://github.com/noir-lang/noir/blob/4a0a672e3658da36eb73d0d4d542208d34e1c7bf/compiler/noirc_frontend/src/elaborator/mod.rs#L458) of passing the expression.

Reply:
That said the [call to deprecated_function](https://github.com/noir-lang/noir/blob/00c186bfaea113799a721d08ce348f703d5153fb/compiler/noirc_frontend/src/elaborator/types.rs#L1441-L1449) shows that's not the only place it's called from, and I agree `lints.rs` would be a nice place for this, I'll move it.

Reply:
Done.